kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemDataset.java

```diff
@@ -399,45 +396,50 @@ public void replace(View<E> update) {
-          FileSystemPartitionView<E> toRemove =
-              (FileSystemPartitionView<E>) partition;
-          Path path = new Path(toRemove.getLocation());
-          removals.add(path);
-          if (partitionListener != null && descriptor.isPartitioned()) {
-            partitionListener.partitionDeleted(
-                namespace, name, toRemove.getRelativeLocation().toString());
+    if (descriptor.isPartitioned()) {
+      // replace leaf partitions one at a time
+      for (PartitionView<E> src : update.getCoveringPartitions()) {
+        if (src instanceof FileSystemPartitionView) {
+          FileSystemPartitionView<E> dest = getPartitionView(
+              ((FileSystemPartitionView<E>) src).getRelativeLocation());
+
+          // The destination partition view may not exist, if the source data was
+          // stored in directories with non-standard names. To account for this,
+          // find all of the directories that should be removed: those that match
+          // the partition constraints.
+          List<Path> removals = Lists.newArrayList();
+          Iterable<PartitionView<E>> existingPartitions = dest
+              .toConstraintsView()
+              .getCoveringPartitions();
+          for (PartitionView<E> partition : existingPartitions) {
+            FileSystemPartitionView<E> toRemove =
+                (FileSystemPartitionView<E>) partition;
+            Path path = new Path(toRemove.getLocation());
+            removals.add(path);
+            if (partitionListener != null && descriptor.isPartitioned()) {
+              partitionListener.partitionDeleted(
+                  namespace, name, toRemove.getRelativeLocation().toString());
+            }
           }
-        }
 
-        // replace the directory all at once
-        FileSystemUtil.replace(fileSystem, directory,
-            new Path(dest.getLocation()), new Path(src.getLocation()),
-            removals);
+          // replace the directory all at once
+          FileSystemUtil.replace(fileSystem, directory,
+              new Path(dest.getLocation()), new Path(src.getLocation()),
+              removals);
 
-        if (partitionListener != null && descriptor.isPartitioned()) {
-          partitionListener.partitionAdded(
-              namespace, name, dest.getRelativeLocation().toString());
-        }
+          if (partitionListener != null && descriptor.isPartitioned()) {
+            partitionListener.partitionAdded(
+                namespace, name, dest.getRelativeLocation().toString());
+          }
 
-      } else {
-        throw new IllegalArgumentException(
-            "Incompatible PartitionView: " + src.getClass().getName());
+        } else {
+          throw new IllegalArgumentException(
+              "Incompatible PartitionView: " + src.getClass().getName());
+        }
       }
+    } else {
+      deleteAll(); // remove all existing files
+      merge((FileSystemDataset<E>) update.getDataset());
```

Comment:
How about making this safer by moving the existing files to a hidden temporary directory then deleting after the merge succeeds?